components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java

```diff
@@ -0,0 +1,203 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.micro.integrator.inbound.endpoint.protocol.grpc;
+
+import com.google.protobuf.Empty;
+import io.grpc.stub.StreamObserver;
+import org.apache.axiom.om.OMElement;
+import org.apache.axiom.om.util.UUIDGenerator;
+import org.apache.axis2.AxisFault;
+import org.apache.axis2.builder.Builder;
+import org.apache.axis2.builder.BuilderUtil;
+import org.apache.axis2.context.MessageContext;
+import org.apache.axis2.transport.TransportUtils;
+import org.apache.commons.io.input.AutoCloseInputStream;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.SynapseConstants;
+import org.apache.synapse.SynapseException;
+import org.apache.synapse.core.SynapseEnvironment;
+import org.apache.synapse.inbound.InboundEndpointConstants;
+import org.apache.synapse.mediators.base.SequenceMediator;
+import org.wso2.micro.integrator.inbound.endpoint.protocol.grpc.util.Event;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+
+import static org.wso2.micro.integrator.inbound.endpoint.protocol.grpc.InboundGrpcConstants.HEADER_MAP_SEQUENCE_PARAMETER_NAME;
+
+/**
+ * Inject gRPC message into the sequence.
+ */
+public class GRPCInjectHandler {
+    private static final Log log = LogFactory.getLog(GRPCInjectHandler.class);
+    private String injectingSeq;
+    private String onErrorSeq;
+    private boolean sequential;
+    private SynapseEnvironment synapseEnvironment;
+
+    /**
+     * constructor initialize parameters and synapseEnvironment
+     *
+     * @param injectingSeq injecting sequence mentioned in the inbound endpoint
+     * @param onErrorSeq error Sequence mentioned in the inbound endpoint
+     * @param sequential is sequential
+     * @param synapseEnvironment The SynapseEnvironment allows access into the the host SOAP engine. It allows
+     *                           the sending of messages, class loader access etc
+     */
+    GRPCInjectHandler(String injectingSeq, String onErrorSeq, boolean sequential,
+                      SynapseEnvironment synapseEnvironment) {
+        this.injectingSeq = injectingSeq;
+        this.onErrorSeq = onErrorSeq;
+        this.sequential = sequential;
+        this.synapseEnvironment = synapseEnvironment;
+    }
+
+    /**
+     * This method will be called when the initiated gRPC call is expecting an response.
+     * This will inject the gRPC message to the to the message sequence as well as the responseObserver .
+     * inject the message.
+     * @param receivedEvent received event from gRPC
+     * @param responseObserver object to be used when sending the response back to the gRPC client
+     */
+    public void invokeProcess(Event receivedEvent, StreamObserver<Event> responseObserver) {
+        try {
+            org.apache.synapse.MessageContext msgCtx = createMessageContext();
+            if (msgCtx != null) {
```

Comment:
Can message context null?

Reply:
No. fixed in https://github.com/wso2/micro-integrator/pull/854/commits/fff5493dd60f14989271869d8d399bb396a56033